tests/Freya.Integration.Tests/Integration.Tests.fs

```diff
@@ -0,0 +1,152 @@
+    let composed = o1.Invoke(o2.Invoke(app))
+
+    let env = invoke composed
+
+    unbox env.["o1"] =? true
+    unbox env.["o2"] =? true
+    env.ContainsKey("Answer") =? true
+    unbox env.["Answer"] =? 42
+
+[<Test>]
+let ``pipeline executes only the first monad if first returns terminate`` () =
+    let app = setLM answerLens 42 |> OwinAppFunc.fromFreya
+    let o1 = modM (fun x -> x.Environment.["o1"] <- true; x) *> halt |> OwinMidFunc.fromFreya
+    let o2 = modM (fun x -> x.Environment.["o2"] <- true; x) *> next |> OwinMidFunc.fromFreya
+    let composed = o1.Invoke(o2.Invoke(app))
+
+    let env = invoke composed
+
+    unbox env.["o1"] =? true
+    unbox env.["o2"] =? false
+    env.ContainsKey("Answer") =? false
+
+[<Test>]
+let ``pipeline executes both monads if first returns next with composed pipeline`` () =
+    let app = setLM answerLens 42 |> OwinAppFunc.fromFreya
+    let o1 = modM (fun x -> x.Environment.["o1"] <- true; x) *> next
+    let o2 = modM (fun x -> x.Environment.["o2"] <- true; x) *> next
+    let pipe = o1 >?= o2 |> OwinMidFunc.fromFreya
+    let composed = pipe.Invoke app
+
+    let env = invoke composed
+
+    unbox env.["o1"] =? true
+    unbox env.["o2"] =? true
+    env.ContainsKey("Answer") =? true
+    unbox env.["Answer"] =? 42
+
+[<Test>]
+let ``pipeline executes only the first monad if first returns terminate with composed pipeline`` () =
+    let app = setLM answerLens 42 |> OwinAppFunc.fromFreya
+    let o1 = modM (fun x -> x.Environment.["o1"] <- true; x) *> halt
+    let o2 = modM (fun x -> x.Environment.["o2"] <- true; x) *> next
+    let pipe = o1 >?= o2 |> OwinMidFunc.fromFreya
+    let composed = pipe.Invoke app
+
+    let env = invoke composed
+
+    unbox env.["o1"] =? true
+    unbox env.["o2"] =? false
+    env.ContainsKey("Answer") =? false
+
+[<Test; Explicit("Implementation has a bug that causes the test to hang.")>]
+let ``MidFunc can be split and used to wrap a pipeline`` () =
+    let stopwatch = System.Diagnostics.Stopwatch()
+    let o1 =
+        modM (fun x ->
+            x.Environment.["o1"] <- true
+            x.Environment.["o1 time"] <- stopwatch.ElapsedMilliseconds
+            x)
+        *> next
+    let o2 =
+        modM (fun x ->
+            x.Environment.["o2"] <- true
+            x.Environment.["o2 time"] <- stopwatch.ElapsedMilliseconds
+            x)
+        *> next
+    let midFunc =
+        OwinMidFunc(fun next ->
+            OwinAppFunc(fun env ->
+                async {
+                    env.["o3"] <- true
+                    env.["o3 time"] <- stopwatch.ElapsedMilliseconds
+                    do! next.Invoke(env).ContinueWith<unit>(fun _ -> ()) |> Async.AwaitTask
+                    env.["o4"] <- true
+                    env.["o4 time"] <- stopwatch.ElapsedMilliseconds
+                } |> Async.StartAsTask :> Task ))
+    let before, after = OwinMidFunc.splitIntoFreya midFunc
```

Comment:
@kolektiv this is an example of what I'm trying to achieve with the `MidFunc`. The above shows a `MidFunc` that performs some actions before and after it executes the `next`. Further tests should show the `MidFunc` halting before calling `next`.